src/lib/firestore-schema.ts

```diff
@@ -0,0 +1,152 @@
+}
+// collection key is portal uid
+type UsersCollection = FSCollection<UserDocument>;
+
+/*
+ * comments
+ *
+ * Subcollection of documents with a denormalized teacher name field with potential synchronization issues.
+ * Comments associated with a document are accessible to any user with access to the document.
+ */
+interface CommentDocument {
+  uid: string;                // user id of author of comment
+  name: string;               // [denormalized] name of user to avoid having to request user data separately
+  createdAt: FSDate;          // timestamp used for sorting
+  tileId?: string;            // empty for document contents
+  content: string;            // plain text for now; potentially html if we need rich text
+}
+// collection key is Firestore-assigned id
+type CommentsCollection = FSCollection<CommentDocument>;
+
+/*
+ * documents
+ *
+ * Subcollection of domain with several metadata fields copied from real-time database (e.g. title, properties)
+ * with potential synchronization issues.
+ * Documents and their comments are accessible to the owning teacher in addition to any network access.
+ * Documents owned by one teacher are accessible to all teachers that share a network with the owner.
+ * Networked access will be mediated by a Firebase function which can check whether the requesting user and
+ * the owning user share a network.
+ */
+interface DocumentDocument {
+  uid: string;
+  type: string;
+  key: string;
+  createdAt: FSDate;
+  title?: string;
+  properties: Record<string, string>;
+  comments: CommentsCollection;
+}
+// collection key is id in Firebase Real-time Database
+type DocumentsCollection = FSCollection<DocumentDocument>;
+
+/*
+ * classes
+ *
+ * Subcollection of domain with several fields copied from portal (e.g. name, teachers) with
+ * potential synchronization issues.
+ * Class resources are accessible to other teachers that share a network with one of the teachers of the class.
+ * Networked access will be mediated by a Firebase function which can check whether the requesting user and
+ * one of the teachers of the class share a network.
+ */
+interface ClassDocument {
+  id: string;                 // portal class id
+  name: string;               // portal class name
+  hash: string;               // portal class hash
+  teachers: string[];         // uids of teachers of class
+}
+// collection key is class id (or class hash?)
+type ClassesCollection = FSCollection<ClassDocument>;
+
+/*
+ * networks
+ *
+ * Subcollection of domain with very little in it. We could try to store a list of users or classes in network
+ * but this would just create synchronization headaches. It seems better to query for those things, at least
+ * initially. I considered making some collections like classes subcollections of networks, but in each case
+ * it seemed like there might be value in being able to access them without going through the network.
+ */
```

Comment:
It is hard to know if it is better for Teachers to have a list of networks. Or Networks to have a list of teachers.  Probably the best option is based on the most common query and optimizing its performance. 

For that reason it might be necessary to denormalize the modeling to optimize the queries and possibly also optimize the access rules. But it seems fine to start like you have and then modify as the queries and access rules take shape.

Reply:
Given that these are going to be edited manually initially, I think finding a teacher and editing a short list of networks whose id's we can choose for human-readability, e.g. `msu-1`, seems easier than finding a network and editing a longer list of `platform_user_id`s.

From an access perspective, I was anticipating that the security rules would allow us to check whether any network of the current teacher overlaps with any network of the teacher(s) that "own" the resource, but as I think about it further security rules allow checking for array containment but not array intersection, so I think the thing to do is to say that a given resource (class, document, etc.) can only be in a single network and to store the network with the resource when it is created in Firestore. I'll make those changes.

Reply:
>~~security rules allow checking for array containment but not array intersection~~

Correction: security rules _do_ actually support list (array) intersection via the [hasAny()](https://firebase.google.com/docs/reference/rules/rules.List#hasAny) function, so we can lookup lists of networks and intersect them.

Reply:
It is fine to start this way, but I think it means the security rules will require a query to lookup the teacher that owns the resource being accessed. This can make the rules slow. Maybe it is easy and fast to get the parent of the document being requested.  Whether it is or not, I think it is OK to start this way and we can optimize later.